### ISSUE TITLE
getTile(x,y,z) method for zoom-able maps

### ISSUE THREAD

**Issue by gszep** [LCSB-BioCore/GigaScatter.jl#13]
We would like a getTile(x,y,z) type method that will eventually be called by some sort of front end such as [OpenLayers](https://openlayers.org/en/latest/examples/xyz.html) that will allow us to zoom in and out of the beautiful SOM Embeddings. Making this work with the Makie plotting ecosystem with both GLMakie and WGLMakie backends would be ideal. Otherwise JSServe.jl can be used to connect to OpenLayers.

**Comment by exaexa** [comment]
Hello!

I'm just adding the links from the e-mails we exchanged, for future reference:

- current functions:  https://makie.juliaplots.org/dev/abstractplotting_api.html#AbstractPlotting.scatter-Tuple  and  https://github.com/JuliaPlots/WGLMakie.jl
- ideal starting point: GeoMaike, https://github.com/JuliaPlots/GeoMakie.jl/issues/67
- example with `getTile`: https://openlayers.org/en/latest/examples/xyz.html

**Comment by gszep** [comment]
probably need to store `embeddingCoordinates =  embedGigaSOM(som,data)` as a quadtree so that we can use the tile index `embeddingCoordinates[x,y,z]` to retrieve the coordinates in that tile.. maybe this package is useful? https://github.com/rdeits/RegionTrees.jl

**Comment by exaexa** [comment]
Yeah, that may help, not sure about the package performance and tree storage overhead though.

For 2D point indexing, the simplest&fastest solution (by far) is usually to have the dataset pre-sorted by z-order integerized coordinates, and stored in full tree in array (like heaps, see e.g. treaps) for cache efficiency -- that's literally few lines of code, and doesn't have any memory overhead, which is a grave concern if we want to render 1G points.

**Comment by gszep** [comment]
useful resource for implementing tile and image coordinates
https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames

**Comment by exaexa** [comment]
Hi Greg! Sorry for the longer delay, stuff has been quite dense.

I've put together a demo (really primitive for now) that does the zoomable embedding trick with the javascript demo you sent:
https://github.com/exaexa/gigascatter-tiles
It runs pretty smoothly for me; there's an example generated "spiral" dataset. let me know if it seems like a step in the "right direction" for you. There are still certainly some TODOs to solve, but nothing really complicated.

**Comment by exaexa** [comment]
PS.
Regarding the original questions:

- **Inplace Rasterisation vs Caching** -- certainly in-place rasterization, it should be fast enough and the browser does the caching for itself.
- **Triggering re-reasterisation** -- if you change the data on the back-end, you'll start getting new colors (or whatever else). One possible problem now is the browser cache (cf. above :] ), but that can be "solved" using some very dirty trick, such as prepending a random nonce to the URL.
